### PR TITLE
fix for ext_authz: config ignored if route does not specify cluster

### LINKED ISSUE #8355: ext_authz: config ignored if route does not specify cluster
As in #8262, the ext_authz filter is a no-op if a route is configured with a direct response action or redirect.

Notably, the filter uses the route/route_entry to obtain a ClusterInfoConstSharedPtr which is used for recording stats. Instead, I think that the filter should produce stats in the context's scope (see the JWT Authn filter, for example) and only produce cluster-specific stats if a cluster is available. In any event, it shouldn't silently skip auth.

A related corner case is that it's possible for the ext_authn response to modify headers and clear the route cache. The re-computed route might send proxy requests to a different cluster (or use a direct response or redirect) but the stats always go to the original cluster.

See also #8250 where the ext_authz filter produced no effect because the header used for cluster selection (via [RouteAction.cluster_header](https://www.envoyproxy.io/docs/envoy/v1.11.1/api-v2/api/v2/route/route.proto.html#envoy-api-field-route-routeaction-cluster-header)) wasn't set yet.

### PR DESCRIPTION
**Pull request by nezdolik**
Signed-off-by: Kateryna Nezdolii <nezdolik@spotify.com>

Description: Currently ext_authz filter acts in noop mode (returns immediately) whenever `cluster info ptr` is null. This could be a case when route is configured with redirect or direct response action. Inside filter this `cluster info ptr` is used to populate stats. Suggested fix is to populate filter related stats within it's own stats object and to set cluster related stats if `cluster info ptr` is non null.
Risk Level: Low/Medium
Testing: Removed 1 test that no longer makes sense and added test.
Docs Changes: TBD
Release Notes: TBD
Fixes #8355

### CODE REVIEW

**Comment by dio** [comment]
Please let me know when it is no longer WIP. Thanks!

**Comment by nezdolik** [comment]
@dio @jmarantz I believe this one is ready for review.

File: source/extensions/filters/http/ext_authz/ext_authz.cc [hunk 1]
```diff
@@ -180,26 +176,30 @@ void Filter::onComplete(Filters::Common::ExtAuthz::ResponsePtr&& response) {
         Http::HeaderMapImpl::appendToHeader(header_to_modify->value(), header.second);
       }
     }
-    config_->incCounter(cluster_->statsScope(), config_->ext_authz_ok_);
+    stats_.ok_.inc();
```
[the remark this answers is not in the document]
Comment (zuercher) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: I asked around and I think we should probably do a deprecation cycle for these stats. So, reinstate the old counters in parallel with the new ones (but with a guard checking if cluster_ is null). Mark them with TODOs and then add release notes noting the change and an entry in in docs/root/intro/deprecated.rst.

[the remark this answers is not in the document]
Comment (nezdolik) [reply source/extensions/filters/http/ext_authz/ext_authz.cc]: Thanks @zuercher , will continue working on this.

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc [hunk 2]
```diff
@@ -203,7 +204,7 @@ TEST_F(HttpFilterTest, ErrorCustomStatusCode) {
     envoy_grpc:
       cluster_name: "ext_authz_server"
   failure_mode_allow: false
-  status_on_error:
```
Comment (nezdolik) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc]: do we need to have spaces in end of lines?

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc [hunk 3]
```diff
@@ -1090,7 +1095,7 @@ TEST_F(HttpFilterTestParam, DeniedResponseWith401) {
   response.status = Filters::Common::ExtAuthz::CheckStatus::Denied;
   response.status_code = Http::Code::Unauthorized;
   request_callbacks_->onComplete(std::make_unique<Filters::Common::ExtAuthz::Response>(response));
-  EXPECT_EQ(1U, filter_callbacks_.clusterInfo()->statsScope().counter("ext_authz.denied").value());
```
Comment (nezdolik) [inline test/extensions/filters/http/ext_authz/ext_authz_test.cc]: missed that test, need to bring back checking of cluster scope stats

File: source/extensions/filters/http/ext_authz/ext_authz.h [hunk 4]
```diff
@@ -99,12 +126,17 @@ class FilterConfig {
   Stats::Scope& scope_;
   Runtime::Loader& runtime_;
   Http::Context& http_context_;
-
+  // todo stop using pool as part of deprecating cluster scope stats.
```
Comment (dio) [inline source/extensions/filters/http/ext_authz/ext_authz.h]: Super nit. I think the convention on placing TODO is `TODO(nezdolik):`

File: test/extensions/filters/http/ext_authz/ext_authz_test.cc (hunk 2 above)
Comment (dio) [reply test/extensions/filters/http/ext_authz/ext_authz_test.cc]: No, I think superfluous space should be cleaned up. Thanks!

File: source/extensions/filters/http/ext_authz/ext_authz.h (hunk 4 above)
Comment (nezdolik) [reply source/extensions/filters/http/ext_authz/ext_authz.h]: @dio, think that `nezdolik` is not allowed word for spell check.

Comment (nezdolik) [reply source/extensions/filters/http/ext_authz/ext_authz.h]: pushed updates TODOs, waiting for spell check results

**Comment by nezdolik** [comment]
@zuercher applied review comments, thanks.

**Comment by mattklein123** [comment]
Sorry can you merge master?

/wait

**Comment by nezdolik** [comment]
@mattklein123 done

**Pull request merged**